lib/Jsrt/JsrtDebugUtils.cpp

```diff
@@ -0,0 +1,465 @@
+
+#include "JsrtPch.h"
+#include "JsrtDebugUtils.h"
+#include "RuntimeDebugPch.h"
+#include "screrror.h"   // For CompileScriptException
+
+void JsrtDebugUtils::AddScriptIdToObject(Js::DynamicObject* object, Js::Utf8SourceInfo* utf8SourceInfo)
+{
+    JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::scriptId, utf8SourceInfo->GetSourceInfoId(), utf8SourceInfo->GetScriptContext());
+
+    Js::Utf8SourceInfo* callerUtf8SourceInfo = utf8SourceInfo->GetCallerUtf8SourceInfo();
+    if (callerUtf8SourceInfo != nullptr)
+    {
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::parentScriptId, callerUtf8SourceInfo->GetSourceInfoId(), callerUtf8SourceInfo->GetScriptContext());
+    }
+}
+
+void JsrtDebugUtils::AddFileNameOrScriptTypeToObject(Js::DynamicObject* object, Js::Utf8SourceInfo* utf8SourceInfo)
+{
+    if (utf8SourceInfo->IsDynamic())
+    {
+        AssertMsg(utf8SourceInfo->GetSourceContextInfo()->url == nullptr, "How come dynamic code have a url?");
+
+        Js::FunctionBody* anyFunctionBody = utf8SourceInfo->GetAnyParsedFunction();
+
+        Assert(anyFunctionBody != nullptr);
+
+        LPCWSTR sourceName = anyFunctionBody->GetSourceName();
+
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::scriptType, sourceName, wcslen(sourceName), utf8SourceInfo->GetScriptContext());
+    }
+    else
+    {
+        // url can be nullptr if JsParseScript/JsRunScript didn't passed any
+        const char16* url = utf8SourceInfo->GetSourceContextInfo()->url == nullptr ? _u("") : utf8SourceInfo->GetSourceContextInfo()->url;
+        JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::fileName, url, wcslen(url), utf8SourceInfo->GetScriptContext());
+    }
+}
+
+void JsrtDebugUtils::AddLineColumnToObject(Js::DynamicObject* object, Js::FunctionBody* functionBody, int byteCodeOffset)
+{
+    if (functionBody != nullptr)
+    {
+        ULONG line = 0;
+        LONG col = 0;
+        if (functionBody->GetLineCharOffset(byteCodeOffset, &line, &col, false))
+        {
+            JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::line, line, functionBody->GetScriptContext());
+            JsrtDebugUtils::AddPropertyToObject(object, JsrtDebugPropertyId::column, col, functionBody->GetScriptContext());
+        }
+    }
+}
+
+void JsrtDebugUtils::AddSourceLengthAndTextToObject(Js::DynamicObject* object, Js::FunctionBody* functionBody, int byteCodeOffset)
+{
+    Js::FunctionBody::StatementMap* statementMap = functionBody->GetEnclosingStatementMapFromByteCode(byteCodeOffset);
+    Assert(statementMap != nullptr);
+
+    LPCUTF8 source = functionBody->GetStartOfDocument(_u("Source for debugging"));
+    size_t startByte = utf8::CharacterIndexToByteIndex(source, functionBody->GetUtf8SourceInfo()->GetCbLength(), (const charcount_t)statementMap->sourceSpan.begin);
+
+    int byteLength = statementMap->sourceSpan.end - statementMap->sourceSpan.begin;
```

Comment:
this is a character difference.

---
In reply to: [62750414](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62750414) [](ancestors = 62750414)